router/proxy/transport.go

```diff
@@ -64,26 +67,55 @@ func (t *transport) setStickyBackend(res *http.Response, originalStickyBackend s
 	}
 }
 
-// always sets the response.Request.URL.Host to the last backend that was
-// connected to.
-func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
+func (t *transport) RoundTrip(ctx context.Context, req *http.Request) (*http.Response, error) {
 	// http.Transport closes the request body on a failed dial, issue #875
-	req.Body = &fakeCloseReadCloser{req.Body}
-	defer req.Body.(*fakeCloseReadCloser).RealClose()
+	fakeBodyCloser := &fakeCloser{req.Body}
+	defer fakeBodyCloser.RealClose()
+
+	cancelc := ctx.Done()
+	reqDone := make(chan struct{})
+	defer close(reqDone)
+
+	req.Body = struct {
+		io.Reader
+		io.Closer
+	}{
+		// This special Reader guards against another subtle race: when the context
+		// is Done() before the httpTransport knows about the request. Details here:
+		// https://go-review.googlesource.com/#/c/2320/5/src/net/http/httputil/reverseproxy.go@120
+		Reader: &runOnFirstRead{
+			Reader: req.Body,
+			fn: func() {
+				go func() {
+					select {
+					case <-cancelc:
+						httpTransport.CancelRequest(req)
+					case <-reqDone:
+					}
```

Comment:
> Even after RoundTrip() is called, you still have a window before the in-flight request has been internally registered by the http.Transport. If CancelRequest() is called during that window, it's a no-op (because the Transport doesn't know about the request).

gotcha, i was failing to understand this window.